### PR TITLE
Add features to bundle On-Premise Gateway related components

### PR DESCRIPTION
**Pull request by npamudika**
### Proposed changes in this pull request

This PR introduces two additional features to bundle On-Premise Gateway related components.
- org.wso2.carbon.apimgt.micro.gateway.default.feature
- org.wso2.carbon.apimgt.micro.gateway.feature

### When should this PR be merged

ASAP

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.tenant.initializer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/tenant/initializer/listener/ServerStartupListener.java [hunk 1]
```diff
@@ -82,41 +82,47 @@ private static void initializeTenant() throws Exception {
...
-        if (CommonUtil.isDomainNameAvailable(tenantDomain)) {
-            tenantInfoBean.setActive(true);
-            tenantInfoBean.setAdmin(email);
-            tenantInfoBean.setAdminPassword(password.toString());
-            tenantInfoBean.setFirstname(TenantInitializationConstants.DEFAULT_FIRST_NAME);
-            tenantInfoBean.setLastname(TenantInitializationConstants.DEFAULT_LAST_NAME);
-            tenantInfoBean.setTenantDomain(tenantDomain);
-            tenantInfoBean.setEmail(email);
+        if (!MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+            String email = MultitenantUtils.getTenantAwareUsername(username);
```
Comment (sanjeewa-malalgoda) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.tenant.initializer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/tenant/initializer/listener/ServerStartupListener.java]: What is the rationale behind this logic. In super tenant space tenant aware user name is email address? Shouldn't we check for enable email user name enabled here before make this decision?

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.tenant.initializer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/tenant/initializer/listener/ServerStartupListener.java [hunk 2]
```diff
@@ -82,41 +82,47 @@ private static void initializeTenant() throws Exception {
...
-                PrivilegedCarbonContext.getThreadLocalCarbonContext()
-                        .setTenantDomain(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME, true);
-                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(MultitenantConstants.SUPER_TENANT_ID);
-
-                tenantMgtAdminService.addTenant(tenantInfoBean);
-                tenantMgtAdminService.activateTenant(tenantDomain);
-            } finally {
-                PrivilegedCarbonContext.endTenantFlow();
+                CommonUtil.validateEmail(email);
+            } catch (Exception e) {
```
Comment (dinusha92) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.tenant.initializer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/tenant/initializer/listener/ServerStartupListener.java]: Shouldn't we throw a specific exception here?

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.tenant.initializer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/tenant/initializer/listener/ServerStartupListener.java [hunk 3]
```diff
@@ -82,41 +82,47 @@ private static void initializeTenant() throws Exception {
...
+                            .setTenantDomain(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME, true);
+                    PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(MultitenantConstants.SUPER_TENANT_ID);
+
+                    tenantMgtAdminService.addTenant(tenantInfoBean);
+                    tenantMgtAdminService.activateTenant(tenantDomain);
+                } finally {
+                    PrivilegedCarbonContext.endTenantFlow();
+                }
+                // Overwriting the char array to clean up password
+                for (int i = 0; i < password.length; i++) {
```
Comment (sanjeewa-malalgoda) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.tenant.initializer/src/main/java/org/wso2/carbon/apimgt/micro/gateway/tenant/initializer/listener/ServerStartupListener.java]: use Arrays.fill(password, '0');

**Comment by npamudika** [comment]
Please refer https://github.com/wso2/product-apim/issues/2805

**Pull request merged**